[PowerToys] Add Command Not Found docs (#4807)
MicrosoftDocs/windows-dev-docs: hub/powertoys/cmd-not-found.md

---
title: PowerToys Command Not Found for Windows
description: A PowerShell 7 module that detects an error thrown by a command and suggests a relevant WinGet package to install, if available.
ms.date: 02/01/2024
ms.topic: article
no-loc: [PowerToys, Windows, Command Not Found, Win]
---

# Command Not Found utility

A PowerShell 7 module that detects an error thrown by a command and suggests a relevant WinGet package to install, if available.

![AlwaysOnTop screenshot.](../images/pt-cmd-not-found.png)

## Requirements

>  - [PowerShell 7](https://learn.microsoft.com/en-us/PowerShell/scripting/install/installing-PowerShell-on-windows?view=PowerShell-7.4)

Reviewer: ```suggestion
 - [PowerShell 7](/PowerShell/scripting/install/installing-PowerShell-on-windows?view=PowerShell-7.4)
```

Is the `?view=PowerShell-7.4` query param needed, or can we let it default to the latest PowerShell docs?

>  - [PowerShell 7](https://learn.microsoft.com/en-us/PowerShell/scripting/install/installing-PowerShell-on-windows?view=PowerShell-7.4)

Reviewer 2: ```suggestion
 - [PowerShell 7](https://learn.microsoft.com/PowerShell/scripting/install/installing-PowerShell-on-windows?view=PowerShell-7.4)
```

 - [PowerShell Microsoft.WinGet.Client module](https://www.powershellgallery.com/packages/Microsoft.WinGet.Client)

## Install the module

> To install the Command Not Found module, navigate to PowerToys settings Command Not Found page and press Install button. On module install, PowerShell 7 experimental features needed for the module to function will be enabled:

Reviewer: ```suggestion
To install the Command Not Found module, navigate to the Command Not Found page in PowerToys settings and select the **Install** button. Once the module installation has completed, PowerShell 7 experimental features needed for the module to function will be enabled:

```

> To install the Command Not Found module, navigate to PowerToys settings Command Not Found page and press Install button. On module install, PowerShell 7 experimental features needed for the module to function will be enabled:

Reviewer 2: ```suggestion
To install the Command Not Found module, go to PowerToys settings Command Not Found page and select **Install**. After installation, PowerShell 7 experimental features needed for the module to function will be enabled:
```

Reviewer: I just left a suggestion for this same block. Can you combine the suggestions into an updated paragraph, @stefansjfw?

Reviewer 2: Haha, we're thinking the same thing! :)

 - PSFeedbackProvider
 - PSCommandNotFoundSuggestion

After that, PowerShell profile file will be appended with following block of PowerShell commands:
```
#34de4b3d-13a8-4540-b76d-b9e8d3851756 PowerToys CommandNotFound module
Import-Module "<powertoys install dir>/WinGetCommandNotFound.psd1"
#34de4b3d-13a8-4540-b76d-b9e8d3851756
```

> Note: If there is no profile file at the time of installing the module, profile file will be created. After installing the module, restart PowerShell session to use the module.

Reviewer: ```suggestion
**Note:** If there is no profile file at the time of installing the module, profile file will be created. After installing the module, restart PowerShell session to use the module.
```
Is bold sufficient, or do you think it should be formatted as a `[!NOTE]`?

Reviewer 2: Honestly and personally, this actually feels more like a normal paragraph to me.

Author: Bold is ok. I'd go with bold to outline this, especially "restart PowerShell" part

> Note: If there is no profile file at the time of installing the module, profile file will be created. After installing the module, restart PowerShell session to use the module.

Reviewer 2: ```suggestion
Note: The profile file will be created if needed. Restart PowerShell session to use the module.
```

## Uninstall the module

> To uninstall the Command Not Found module, navigate to PowerToys settings Command Not Found page and press Uninstall button. On module uninstall, previously added block of commands will be removed from PowerShell profile file. 

Reviewer: ```suggestion
To uninstall the Command Not Found module, navigate to the Command Not Found page in PowerToys settings and select the **Uninstall** button. Once the module uninstall has completed, the block of commands previously added will be removed from the PowerShell profile file. 
```

> To uninstall the Command Not Found module, navigate to PowerToys settings Command Not Found page and press Uninstall button. On module uninstall, previously added block of commands will be removed from PowerShell profile file. 

Reviewer 2: ```suggestion
To uninstall the Command Not Found module, go to PowerToys settings Command Not Found page and select **Uninstall**. After uninstallation, previously added block of commands will be removed from PowerShell profile file. 
```

---
title: PowerToys Command Not Found for Windows
description: A PowerShell 7 module that detects an error thrown by a command and suggests a relevant WinGet package to install, if available.
ms.date: 02/01/2024
ms.topic: article
no-loc: [PowerToys, Windows, Command Not Found, Win]
---

# Command Not Found utility

> A PowerShell 7 module that detects an error thrown by a command and suggests a relevant WinGet package to install, if available.

Reviewer 3: ```suggestion
A PowerShell 7 module that detects command-line errors and suggests a relevant WinGet package to install, if available.
```

This is just a small suggestion but I'm not sure about 'thrown' as a verb to describe errors. I know it's used for [exceptions in these docs](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/exception-throwing) but I can't see it in many other Microsoft docs.

![Command Not Found screenshot.](../images/pt-cmd-not-found.png)

> ## Requirements

Reviewer 2: ```suggestion
## Requirements

```

 - [PowerShell 7](/PowerShell/scripting/install/installing-PowerShell-on-windows)
 - [PowerShell Microsoft.WinGet.Client module](https://www.powershellgallery.com/packages/Microsoft.WinGet.Client)

## Install the module

To install the Command Not Found module, navigate to the Command Not Found page in PowerToys settings and select the **Install** button. Once the module installation has completed, PowerShell 7 experimental features needed for the module to function will be enabled:

 - PSFeedbackProvider
 - PSCommandNotFoundSuggestion

After that, PowerShell profile file will be appended with following block of PowerShell commands:

> ```

Reviewer 2: What's the language for this code snippet? psh?

Author: yes

Reviewer 2: ```suggestion

```psh
```
